### PR TITLE
CIRC-1767 Add additionalInfo token to the fee/fine notice context

### PR DESCRIPTION
**Pull request by alexanderkurash**
Resolves [CIRC-1767](https://issues.folio.org/browse/CIRC-1767)

## Purpose
Users need to be able to include the "Additional information for patron" into their fee/fine related patron notice templates as feeCharge.additionalInformation

## Approach
Extend existing implementation

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/Account.java [hunk 1]
```diff
@@ -197,4 +198,9 @@ private int compareByDateAction(FeeFineAction left, FeeFineAction right) {
     return compareToMillis(left.getDateAction(), right.getDateAction());
   }
 
+  @Override
+  public String toString() {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/Account.java:202]: Use lombok

File: src/main/java/org/folio/circulation/domain/FeeFineAction.java [hunk 2]
```diff
@@ -70,4 +71,10 @@ public static FeeFineAction from(JsonObject representation) {
   public String getPaymentMethod() {
     return representation.getString("paymentMethod");
   }
+
+  @Override
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/FeeFineAction.java:75]: Use lombok

File: src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java [hunk 3]
```diff
@@ -49,6 +61,27 @@ public CompletableFuture<Result<FeeFineAction>> findById(String id) {
...
+
+    Result<CqlQuery> query = CqlQuery.exactMatch("accountId", account.getId())
+      .combine(exactMatch("typeAction", account.getFeeFineType()), CqlQuery::and);
+
+    return new CqlQueryFinder<>(feeFineActionsStorageClient, "feefineactions", identity())
+      .findByQuery(query, PageLimit.one())
+      .thenApply(mapResult(records -> records.mapRecords(FeeFineAction::from)))
+      .thenApply(mapResult(MultipleRecords::getRecords))
+      .thenApply(mapResult(records -> {
+        log.info("findChargeActionForAccount:: found {} fee/fine actions for account {}",
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java:79]: debug

File: src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java [hunk 4]
```diff
@@ -49,6 +61,27 @@ public CompletableFuture<Result<FeeFineAction>> findById(String id) {
...
+          records.size(), account);
+        return records.stream().findFirst().orElse(null);
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java:81]: firstOrNull

File: src/main/java/org/folio/circulation/domain/notice/TemplateContextUtil.java [hunk 5]
```diff
@@ -275,21 +279,24 @@ private static JsonObject createLoanContext(Loan loan) {
...
-    return createFeeFineNoticeContext(account, loan)
-      .put(FEE_ACTION, createFeeActionContext(feeFineAction));
+    return createFeeFineChargeNoticeContext(account, loan, chargeAction)
+      .put(FEE_ACTION, createFeeActionContext(currentAction));
   }
 
-  private static JsonObject createFeeChargeContext(Account account) {
-    JsonObject context = new JsonObject();
+  private static JsonObject createFeeChargeContext(Account account, FeeFineAction chargeAction) {
+    log.debug("createFeeChargeContext:: params account{}, chargeAction={}", account, chargeAction);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/TemplateContextUtil.java:297]:
```suggestion
    log.debug("createFeeChargeContext:: params account={}, chargeAction={}", account, chargeAction);
```

File: src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java [hunk 6]
```diff
@@ -49,6 +61,22 @@ public CompletableFuture<Result<FeeFineAction>> findById(String id) {
       .fetch(id);
   }
 
+  public CompletableFuture<Result<FeeFineAction>> findChargeActionForAccount(Account account) {
+    log.debug("findChargeActionForAccount:: params account={}", account);
+
+    if (isNull(account)) {
+      return ofAsync(() -> null);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java:68]: emptyAsync?

File: src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java [hunk 7]
```diff
@@ -49,6 +61,22 @@ public CompletableFuture<Result<FeeFineAction>> findById(String id) {
...
+    }
+
+    Result<CqlQuery> query = CqlQuery.exactMatch("accountId", account.getId())
+      .combine(exactMatch("typeAction", account.getFeeFineType()), CqlQuery::and);
+
+    return new CqlQueryFinder<>(feeFineActionsStorageClient, "feefineactions", identity())
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java:74]:
```suggestion
    return new CqlQueryFinder<>(feeFineActionsStorageClient, "feefineactions", FeeFineAction::from)
```

File: src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java [hunk 8]
```diff
@@ -49,6 +61,22 @@ public CompletableFuture<Result<FeeFineAction>> findById(String id) {
...
+      .findByQuery(query, PageLimit.one())
+      .thenApply(mapResult(records -> records.mapRecords(FeeFineAction::from)))
+      .thenApply(mapResult(MultipleRecords::firstOrNull));
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/infrastructure/storage/feesandfines/FeeFineActionRepository.java:77]:
```suggestion
    return new CqlQueryFinder<>(feeFineActionsStorageClient, "feefineactions", FeeFineAction::from)
      .findByQuery(query, PageLimit.one())
      .thenApply(mapResult(MultipleRecords::firstOrNull));
```

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**